llvm/lib/Target/AIE/AIECombinerHelper.cpp

```diff
@@ -76,6 +76,34 @@ bool isUseOf(const MachineInstr &MI, const MachineInstr &Use) {
   return false;
 }
 
+///  Check for dead \a InBetweenMI MI and copy-like instructions that can be
+///  coalesced once \a MemI and \a Dest are combined.
+bool isNonCoalesceableUseOf(const MachineInstr &MemI,
+                            const MachineInstr &InBetweenMI,
+                            const MachineInstr &Dest,
+                            const MachineRegisterInfo &MRI) {
+  bool IsCopyBetween = false;
+
+  if (isTriviallyDead(InBetweenMI, MRI))
+    return false;
+
+  // We can delay an instruction after a copy, if the copy just
+  // connects MemI and Dest. After combining, this copy will be dead.
+  if (InBetweenMI.isCopy() &&
+      MRI.hasOneNonDBGUse(InBetweenMI.getOperand(1).getReg()) &&
+      MRI.hasOneNonDBGUse(InBetweenMI.getOperand(0).getReg())) {
+    const MachineInstr *CopyOrignMI =
+        MRI.getVRegDef(InBetweenMI.getOperand(1).getReg());
+    const MachineInstr *CopyDestMI =
+        &*MRI.use_instr_nodbg_begin(InBetweenMI.getOperand(0).getReg());
```

Comment:
We verify that the bitcast has a single use, but not the intermediate copies. Maybe we should just handle `TargetOpcode::COPY` specifically as well?

Reply:
It makes sense. I update the PR and included a test for copy with no single use.